### ISSUE TITLE
Arbitrary crash of sound mixing

### ISSUE THREAD

**Issue by gurkanctn** [OneLoneCoder/olcPixelGameEngine#37]
I'm using the OLC Pixel Game Engine, with sounds, in my game [Altay Tank Game](https://epicos.itch.io/altay-tank-game).

Recently, with game v1.0, I've started to receive some bug reports saying that the game sometimes halts (stops responding) without a total crash. I've also come across a similar situation when I was running through Visual Studio, Release build.

In my experience the following lines were highlighted (specifically _add to the mix_ line):

````
else
	{
		// Calculate sample position
		s.nSamplePosition += (long)((float)vecAudioSamples[s.nAudioSampleID - 1].wavHeader.nSamplesPerSec * fTimeStep);
		// If sample position is valid add to the mix
		if (s.nSamplePosition < vecAudioSamples[s.nAudioSampleID - 1].nSamples)
			fMixerSample += vecAudioSamples[s.nAudioSampleID - 1].fSample[(s.nSamplePosition * vecAudioSamples[s.nAudioSampleID - 1].nChannels) + nChannel];
	else

````

Just wondering if this is a Windows sound mixing issue, and I'd like to know if there's a Best Practice I might be missing. I'm roughly guessing that this halt might occur when several sounds are being mixed and played simultaneously.

My game source code is available on Github: [https://github.com/gurkanctn/AltayTankGame/blob/master/Altay.cpp](https://github.com/gurkanctn/AltayTankGame/blob/master/Altay.cpp)

Thanks for any tips and suggestions or a fix.

Greetings,

**Comment by OneLoneCoder** [comment]
Hi Gurkan,
Apologies for the late response, I don't check github very frequently. The code sample you have provided does not have the potential to lock up under normal circumstances. This suggests that you could be trying to play a sample with an invalid index. At run time, it has no way of knowing if the ID is invalid, thus playing with memory it shouldn't be.
I dont think this will be a windows mixing issue, but it is difficult to isolate the cause of the lock. There is potential that for trouble if you are manipulating the vector of samples whilst the sound engine is playing.
The sound engine is due for a bit of an overhaul and some documentation soon enough, so thanks for the reminder!
Jx9

**Comment by gurkanctn** [comment]
Hello javidx9, thanks for the reply. I'm not in a hurry at all, life is at a fast pace everywhere. :)

> There is potential that for trouble if you are manipulating the vector of samples whilst the sound engine is playing.

Actually, I'm not manipulating the vector of samples at all.

After some debugging, I've noticed that one of my new sound fx files was broken. It cannot be loaded successfully to Sound engine, but can be played on Windows Media Player. I've fixed my problem already (I'm playing a working file).

I've also done some performance test to play SEVERAL SOUNDS SIMULTANEOUSLY and it successfully plays tens of instances of my three sound fx without crashing.

Lesson learned: I'll add some code to validate that all sound files (and sprites) are successfully loaded into the memory.

I guess I was assuming that the PGE would be checking file/data validity and would crash right upon OnUserCreate(). But, no problems! I love the PGE. :)
